build_tools/cmake/iree_test_matrix.cmake

```diff
@@ -0,0 +1,24 @@
+# Copyright 2021 The IREE Authors
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+include(CMakeParseArguments)
+
+function(iree_gen_test_matrix)
+  set(_test_matrix_yaml "${CMAKE_CURRENT_SOURCE_DIR}/test_matrix.yaml")
+  set(_generator_script "${PROJECT_SOURCE_DIR}/build_tools/testing/gen_test_matrix.py")
+  message(STATUS "Generating tests for ${CMAKE_CURRENT_SOURCE_DIR}")
```

Comment:
This is happening at configure time, right?

* About how long does it take? Is reconfiguring cheaper than a clean build?
* Does it clear the generated directory before adding new files? Can orphaned files be created and not cleaned up?

Reply:
I was originally running it interactively while developing and I boosted it into cmake when I saw that it is basically instantaneous and a better UX to do it that way. The script cleans up any generated tests that are stale on the next configure, so I don't think it will have a problem with orphans.

It could be optimized in various ways, but right now at "imperceptible" seems ok. Probably becomes perceptible on windows, if history is a guide :)

Reply:
Ok, I don't mind if it's imperceptible :)

Putting the tests and their generation behind a CMake flag is an easy step to take to step around any rough edges for development in other parts of the tree. If we get to the point where we have tests for multiple frontend frameworks generated in a large matrix (e.g. 1000s of files) then we can measure and optimize as needed.